staging/src/k8s.io/mount-utils/mount.go

```diff
@@ -147,6 +147,23 @@ func NewMountError(mountErrorValue MountErrorType, format string, args ...interf
 type SafeFormatAndMount struct {
 	Interface
 	Exec utilexec.Interface
+
+	formatSem     chan any
+	formatTimeout time.Duration
+}
+
+// SetMaxConcurrentFormat sets the maximum number of concurrent format
+// operations executed by the mounter. The timeout controls the maximum
+// duration of a format operation before it is treated as complete. A max
+// concurrent value of less than 1 sets concurrency to unlimited.
+// This function assumes that no format operations are currently in flight.
+func (mounter *SafeFormatAndMount) SetMaxConcurrentFormat(n int, timeout time.Duration) {
```

Comment:
is it expected that no in-tree code is making use of this?

Reply:
Nothing explicitly prevents it. I'm not sure I understand the implications. Could you please elaborate?

Reply:
the PR refers to OOMs and problems with highly concurrent formatting... is that a general issue that in-tree callers should be preventing by limiting the concurrency of how they use this mounter?

Reply:
Here's my take on the context, hope it's helpful:

SafeFormatAndMount is used by in-tree code as well as other cloud provider csi drivers (eg, [ebs](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/master/pkg/mounter/safe_mounter_unix.go)).

The kubelet runs this formatting as part of the volume stage operation, and typically these run in their own thread. Concurrent mkfs hasn't been a problem for the kubelet, because IIUC there's no memory limit for the kubelet container. But it's a problem for CSI drivers that have a limit. In particular with many (10s) of mkfs's happening, the driver will crash-loop and not make any progress.

I don't know how, eg, EBS is deployed on EKS, maybe they'll run into this problem, maybe they run their driver with no limit, or maybe they let users modify the deployment to bump the limit if they have problems in practice.

Currently, if a mkfs gets stuck, it will block its operation (and so block the starting of the pod using the volume), but this won't block other volumes/pods from scheduling. That's nice semantics to maintain, which is why there's a timeout. If it's rare for mkfs to get stuck then pods will be able to make progress scheduling on a node and dealing with one stuck pod is tractable.

It seems this might be a useful change for other CSI drivers to make use of.

in-tree code could make use of this---but all the in-tree drivers are deprecated so I don't expect we're going to update them (especially since this doesn't seem to have been a problem in practice).

Reply:
that context is helpful, thanks